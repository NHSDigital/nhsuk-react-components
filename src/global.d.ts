declare module 'nhsuk-frontend/packages/components/header/header.js';
declare module 'nhsuk-frontend/packages/components/checkboxes/checkboxes.js';
declare module 'nhsuk-frontend/packages/components/character-count/character-count.js';
declare module 'nhsuk-frontend/packages/components/tabs/tabs.js';
