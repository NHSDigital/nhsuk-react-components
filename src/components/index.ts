export * from './content-presentation/index.js';
export * from './form-elements/index.js';
export * from './layout/index.js';
export * from './navigation/index.js';
export * from './typography/index.js';
export * from './utils/index.js';
