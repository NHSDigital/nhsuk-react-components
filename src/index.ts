export * from './components';
export * from './patterns';
