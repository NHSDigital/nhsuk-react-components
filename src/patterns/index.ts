export * from './nav-a-z/index.js';
export * from './review-date/index.js';
