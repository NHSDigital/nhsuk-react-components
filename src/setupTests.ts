import { TextDecoder, TextEncoder } from 'util';
import '@testing-library/jest-dom';
import { outdent } from 'outdent';

/**
 * Polyfill TextEncoder/TextDecoder for ReactDOM
 *
 * @see {@link https://github.com/jsdom/jsdom/issues/1695}
 */
Object.defineProperties(globalThis, {
  TextEncoder: {
    value: TextEncoder,
    configurable: true,
    enumerable: false,
    writable: true,
  },
  TextDecoder: {
    value: TextDecoder,
    configurable: true,
    enumerable: false,
    writable: true,
  },
});

/**
 * Polyfill Element methods for NHS.UK frontend error summary
 *
 * @see {@link https://github.com/jsdom/jsdom/issues/1695}
 */
Object.defineProperties(Element.prototype, {
  scrollIntoView: { value: jest.fn() },
});

/**
 * Polyfill `matchMedia()` for NHS.UK frontend tabs
 */
Object.defineProperty(globalThis, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: true,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    addEventListener: jest.fn(),
    removeListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

beforeEach(() => {
  const stylesheet = document.createElement('style');

  stylesheet.innerHTML = outdent`
    :root {
      --nhsuk-breakpoint-mobile: 20rem;
      --nhsuk-breakpoint-tablet: 40.0625rem;
      --nhsuk-breakpoint-desktop: 48.0625rem;
      --nhsuk-breakpoint-large-desktop: 61.875rem;
    }
  `;

  // Add styles for NHS.UK frontend checks
  document.head.appendChild(stylesheet);

  // Flag NHS.UK frontend as supported
  document.body.classList.add('nhsuk-frontend-supported');
});
