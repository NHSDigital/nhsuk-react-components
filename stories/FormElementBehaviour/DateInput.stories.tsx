import React, { useState } from 'react';
import { DateInput } from '../../src';
import { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof DateInput> = {
  title: 'FormBehaviour/DateInput',
  component: DateInput,
};
export default meta;
type Story = StoryObj<typeof DateInput>;

export const Standard: Story = {
  render: (args) => (
    <div style={{ padding: 20 }}>
      <h2>Scenario: onChange and onInput handlers are bound without any other props</h2>
      <h5>Expected Behaviour</h5>
      <ul className="nhsuk-hint">
        <li>OnChange Handlers are fired using the generated IDs and Names</li>
        <li>The value is passed through</li>
      </ul>
      <h5>Component</h5>
      <DateInput
        onChange={(e) => console.log(e.target.value)}
        hint="Test hint"
        label="Test label"
      />
    </div>
  ),
};

export const StandardWithError: Story = {
  render: (args) => (
    <div style={{ padding: 20 }}>
      <h2>Scenario: onChange and onInput handlers are bound without any other props</h2>
      <h5>Expected Behaviour</h5>
      <ul className="nhsuk-hint">
        <li>OnChange Handlers are fired using the generated IDs and Names</li>
        <li>The value is passed through</li>
      </ul>
      <h5>Component</h5>
      <DateInput
        onChange={(e) => console.log(e.target.value)}
        error="Test Error"
        hint="Test hint"
        label="Test label"
      />
      <h5>Component with specific field errors</h5>

      <DateInput
        onChange={(e) => console.log(e.target.value)}
        error="Test Error"
        hint="Test hint"
        label="Test label"
      >
        <DateInput.Day error={false} />
        <DateInput.Month />
        <DateInput.Year />
      </DateInput>
    </div>
  ),
};

export const PrePopulatedIndividualComponents: Story = {
  render: (args) => {
    const defaultValue = { day: '20', month: '09', year: '1996' };
    return (
      <div style={{ padding: 20 }}>
        <h5>Component</h5>
        <DateInput hint="Test hint" label="Test label">
          <DateInput.Day defaultValue={defaultValue.day} />
          <DateInput.Month defaultValue={defaultValue.month} />
          <DateInput.Year defaultValue={defaultValue.year} />
        </DateInput>
      </div>
    );
  },
};

export const PrePopulatedWrapper: Story = {
  render: (args) => {
    const defaultValue = { day: '20', month: '09', year: '1996' };
    return (
      <div style={{ padding: 20 }}>
        <h5>Component</h5>
        <DateInput hint="Test hint" label="Test label" defaultValue={defaultValue} />
      </div>
    );
  },
};

export const ControlledElementIndividualComponents: Story = {
  render: (args) => {
    const value = { day: '20', month: '09', year: '1996' };
    return (
      <div style={{ padding: 20 }}>
        <h5>Component</h5>
        <DateInput hint="Test hint" label="Test label">
          <DateInput.Day value={value.day} />
          <DateInput.Month value={value.month} />
          <DateInput.Year value={value.year} />
        </DateInput>
      </div>
    );
  },
};

export const ControlledElementWrapper: Story = {
  render: (args) => {
    const value = { day: '20', month: '09', year: '1996' };
    return (
      <div style={{ padding: 20 }}>
        <h5>Component</h5>
        <DateInput hint="Test hint" label="Test label" value={value} />
      </div>
    );
  },
};

export const ChangeableControlledElement: Story = {
  render: function ChangeableControlledElementRender() {
    const [value, setValue] = useState({ day: '20', month: '09', year: '1996' });

    return (
      <div style={{ padding: 20 }}>
        <h5>Component</h5>
        <DateInput
          hint="Test hint"
          label="Test label"
          value={value}
          onChange={(e) => setValue(e.currentTarget.value)}
        />
      </div>
    );
  },
};
