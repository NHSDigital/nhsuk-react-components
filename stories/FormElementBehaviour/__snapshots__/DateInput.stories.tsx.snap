// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBehaviour/DateInput ChangeableControlledElement smoke-test 1`] = `
<div style="padding: 20px;">
  <h5>
    Component
  </h5>
  <div class="nhsuk-form-group">
    <label class="nhsuk-label"
           id="dateinput_6xm6p--label"
           for="dateinput_6xm6p"
    >
      Test label
    </label>
    <div class="nhsuk-hint"
         id="dateinput_6xm6p--hint"
    >
      Test hint
    </div>
    <div class="nhsuk-date-input"
         aria-describedby="dateinput_6xm6p--hint"
         aria-labelledby="dateinput_6xm6p--label"
         id="dateinput_6xm6p"
    >
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_6xm6p-day--label"
                 for="dateinput_6xm6p-day"
          >
            Day
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_6xm6p-day"
                 aria-labelledby="dateinput_6xm6p-day--label"
                 name="dateinput_6xm6p-day"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="20"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_6xm6p-month--label"
                 for="dateinput_6xm6p-month"
          >
            Month
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_6xm6p-month"
                 aria-labelledby="dateinput_6xm6p-month--label"
                 name="dateinput_6xm6p-month"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="09"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_6xm6p-year--label"
                 for="dateinput_6xm6p-year"
          >
            Year
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
                 id="dateinput_6xm6p-year"
                 aria-labelledby="dateinput_6xm6p-year--label"
                 name="dateinput_6xm6p-year"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="1996"
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormBehaviour/DateInput ControlledElementIndividualComponents smoke-test 1`] = `
<div style="padding: 20px;">
  <h5>
    Component
  </h5>
  <div class="nhsuk-form-group">
    <label class="nhsuk-label"
           id="dateinput_0amxm--label"
           for="dateinput_0amxm"
    >
      Test label
    </label>
    <div class="nhsuk-hint"
         id="dateinput_0amxm--hint"
    >
      Test hint
    </div>
    <div class="nhsuk-date-input"
         aria-describedby="dateinput_0amxm--hint"
         aria-labelledby="dateinput_0amxm--label"
         id="dateinput_0amxm"
    >
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_0amxm-day--label"
                 for="dateinput_0amxm-day"
          >
            Day
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_0amxm-day"
                 aria-labelledby="dateinput_0amxm-day--label"
                 name="dateinput_0amxm-day"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="20"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_0amxm-month--label"
                 for="dateinput_0amxm-month"
          >
            Month
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_0amxm-month"
                 aria-labelledby="dateinput_0amxm-month--label"
                 name="dateinput_0amxm-month"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="09"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_0amxm-year--label"
                 for="dateinput_0amxm-year"
          >
            Year
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
                 id="dateinput_0amxm-year"
                 aria-labelledby="dateinput_0amxm-year--label"
                 name="dateinput_0amxm-year"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="1996"
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormBehaviour/DateInput ControlledElementWrapper smoke-test 1`] = `
<div style="padding: 20px;">
  <h5>
    Component
  </h5>
  <div class="nhsuk-form-group">
    <label class="nhsuk-label"
           id="dateinput_42lx9--label"
           for="dateinput_42lx9"
    >
      Test label
    </label>
    <div class="nhsuk-hint"
         id="dateinput_42lx9--hint"
    >
      Test hint
    </div>
    <div class="nhsuk-date-input"
         aria-describedby="dateinput_42lx9--hint"
         aria-labelledby="dateinput_42lx9--label"
         id="dateinput_42lx9"
    >
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_42lx9-day--label"
                 for="dateinput_42lx9-day"
          >
            Day
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_42lx9-day"
                 aria-labelledby="dateinput_42lx9-day--label"
                 name="dateinput_42lx9-day"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="20"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_42lx9-month--label"
                 for="dateinput_42lx9-month"
          >
            Month
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_42lx9-month"
                 aria-labelledby="dateinput_42lx9-month--label"
                 name="dateinput_42lx9-month"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="09"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_42lx9-year--label"
                 for="dateinput_42lx9-year"
          >
            Year
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
                 id="dateinput_42lx9-year"
                 aria-labelledby="dateinput_42lx9-year--label"
                 name="dateinput_42lx9-year"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="1996"
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormBehaviour/DateInput PrePopulatedIndividualComponents smoke-test 1`] = `
<div style="padding: 20px;">
  <h5>
    Component
  </h5>
  <div class="nhsuk-form-group">
    <label class="nhsuk-label"
           id="dateinput_lsnrd--label"
           for="dateinput_lsnrd"
    >
      Test label
    </label>
    <div class="nhsuk-hint"
         id="dateinput_lsnrd--hint"
    >
      Test hint
    </div>
    <div class="nhsuk-date-input"
         aria-describedby="dateinput_lsnrd--hint"
         aria-labelledby="dateinput_lsnrd--label"
         id="dateinput_lsnrd"
    >
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_lsnrd-day--label"
                 for="dateinput_lsnrd-day"
          >
            Day
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_lsnrd-day"
                 aria-labelledby="dateinput_lsnrd-day--label"
                 name="dateinput_lsnrd-day"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="20"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_lsnrd-month--label"
                 for="dateinput_lsnrd-month"
          >
            Month
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_lsnrd-month"
                 aria-labelledby="dateinput_lsnrd-month--label"
                 name="dateinput_lsnrd-month"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="09"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_lsnrd-year--label"
                 for="dateinput_lsnrd-year"
          >
            Year
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
                 id="dateinput_lsnrd-year"
                 aria-labelledby="dateinput_lsnrd-year--label"
                 name="dateinput_lsnrd-year"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="1996"
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormBehaviour/DateInput PrePopulatedWrapper smoke-test 1`] = `
<div style="padding: 20px;">
  <h5>
    Component
  </h5>
  <div class="nhsuk-form-group">
    <label class="nhsuk-label"
           id="dateinput_skrkf--label"
           for="dateinput_skrkf"
    >
      Test label
    </label>
    <div class="nhsuk-hint"
         id="dateinput_skrkf--hint"
    >
      Test hint
    </div>
    <div class="nhsuk-date-input"
         aria-describedby="dateinput_skrkf--hint"
         aria-labelledby="dateinput_skrkf--label"
         id="dateinput_skrkf"
    >
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_skrkf-day--label"
                 for="dateinput_skrkf-day"
          >
            Day
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_skrkf-day"
                 aria-labelledby="dateinput_skrkf-day--label"
                 name="dateinput_skrkf-day"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="20"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_skrkf-month--label"
                 for="dateinput_skrkf-month"
          >
            Month
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_skrkf-month"
                 aria-labelledby="dateinput_skrkf-month--label"
                 name="dateinput_skrkf-month"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="09"
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_skrkf-year--label"
                 for="dateinput_skrkf-year"
          >
            Year
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
                 id="dateinput_skrkf-year"
                 aria-labelledby="dateinput_skrkf-year--label"
                 name="dateinput_skrkf-year"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value="1996"
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormBehaviour/DateInput Standard smoke-test 1`] = `
<div style="padding: 20px;">
  <h2>
    Scenario: onChange and onInput handlers are bound without any other props
  </h2>
  <h5>
    Expected Behaviour
  </h5>
  <ul class="nhsuk-hint">
    <li>
      OnChange Handlers are fired using the generated IDs and Names
    </li>
    <li>
      The value is passed through
    </li>
  </ul>
  <h5>
    Component
  </h5>
  <div class="nhsuk-form-group">
    <label class="nhsuk-label"
           id="dateinput_qy0p0--label"
           for="dateinput_qy0p0"
    >
      Test label
    </label>
    <div class="nhsuk-hint"
         id="dateinput_qy0p0--hint"
    >
      Test hint
    </div>
    <div class="nhsuk-date-input"
         aria-describedby="dateinput_qy0p0--hint"
         aria-labelledby="dateinput_qy0p0--label"
         id="dateinput_qy0p0"
    >
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_qy0p0-day--label"
                 for="dateinput_qy0p0-day"
          >
            Day
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_qy0p0-day"
                 aria-labelledby="dateinput_qy0p0-day--label"
                 name="dateinput_qy0p0-day"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_qy0p0-month--label"
                 for="dateinput_qy0p0-month"
          >
            Month
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_qy0p0-month"
                 aria-labelledby="dateinput_qy0p0-month--label"
                 name="dateinput_qy0p0-month"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_qy0p0-year--label"
                 for="dateinput_qy0p0-year"
          >
            Year
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
                 id="dateinput_qy0p0-year"
                 aria-labelledby="dateinput_qy0p0-year--label"
                 name="dateinput_qy0p0-year"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormBehaviour/DateInput StandardWithError smoke-test 1`] = `
<div style="padding: 20px;">
  <h2>
    Scenario: onChange and onInput handlers are bound without any other props
  </h2>
  <h5>
    Expected Behaviour
  </h5>
  <ul class="nhsuk-hint">
    <li>
      OnChange Handlers are fired using the generated IDs and Names
    </li>
    <li>
      The value is passed through
    </li>
  </ul>
  <h5>
    Component
  </h5>
  <div class="nhsuk-form-group nhsuk-form-group--error">
    <label class="nhsuk-label"
           id="dateinput_ltr3j--label"
           for="dateinput_ltr3j"
    >
      Test label
    </label>
    <div class="nhsuk-hint"
         id="dateinput_ltr3j--hint"
    >
      Test hint
    </div>
    <span class="nhsuk-error-message"
          id="dateinput_ltr3j--error-message"
          role="alert"
    >
      <span class="nhsuk-u-visually-hidden">
        Error:
      </span>
      Test Error
    </span>
    <div class="nhsuk-date-input"
         aria-describedby="dateinput_ltr3j--hint"
         aria-labelledby="dateinput_ltr3j--label"
         id="dateinput_ltr3j"
    >
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_ltr3j-day--label"
                 for="dateinput_ltr3j-day"
          >
            Day
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2 nhsuk-input--error"
                 id="dateinput_ltr3j-day"
                 aria-labelledby="dateinput_ltr3j-day--label"
                 name="dateinput_ltr3j-day"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_ltr3j-month--label"
                 for="dateinput_ltr3j-month"
          >
            Month
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2 nhsuk-input--error"
                 id="dateinput_ltr3j-month"
                 aria-labelledby="dateinput_ltr3j-month--label"
                 name="dateinput_ltr3j-month"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_ltr3j-year--label"
                 for="dateinput_ltr3j-year"
          >
            Year
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4 nhsuk-input--error"
                 id="dateinput_ltr3j-year"
                 aria-labelledby="dateinput_ltr3j-year--label"
                 name="dateinput_ltr3j-year"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
    </div>
  </div>
  <h5>
    Component with specific field errors
  </h5>
  <div class="nhsuk-form-group nhsuk-form-group--error">
    <label class="nhsuk-label"
           id="dateinput_4hpsm--label"
           for="dateinput_4hpsm"
    >
      Test label
    </label>
    <div class="nhsuk-hint"
         id="dateinput_4hpsm--hint"
    >
      Test hint
    </div>
    <span class="nhsuk-error-message"
          id="dateinput_4hpsm--error-message"
          role="alert"
    >
      <span class="nhsuk-u-visually-hidden">
        Error:
      </span>
      Test Error
    </span>
    <div class="nhsuk-date-input"
         aria-describedby="dateinput_4hpsm--hint"
         aria-labelledby="dateinput_4hpsm--label"
         id="dateinput_4hpsm"
    >
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_4hpsm-day--label"
                 for="dateinput_4hpsm-day"
          >
            Day
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                 id="dateinput_4hpsm-day"
                 aria-labelledby="dateinput_4hpsm-day--label"
                 name="dateinput_4hpsm-day"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_4hpsm-month--label"
                 for="dateinput_4hpsm-month"
          >
            Month
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2 nhsuk-input--error"
                 id="dateinput_4hpsm-month"
                 aria-labelledby="dateinput_4hpsm-month--label"
                 name="dateinput_4hpsm-month"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
      <div class="nhsuk-date-input__item">
        <div class="nhsuk-form-group">
          <label class="nhsuk-label nhsuk-date-input__label"
                 id="dateinput_4hpsm-year--label"
                 for="dateinput_4hpsm-year"
          >
            Year
          </label>
          <input class="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4 nhsuk-input--error"
                 id="dateinput_4hpsm-year"
                 aria-labelledby="dateinput_4hpsm-year--label"
                 name="dateinput_4hpsm-year"
                 pattern="[0-9]*"
                 inputmode="numeric"
                 type="text"
                 value
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;
